src/lib/scheduler/operator_task.cpp

```diff
@@ -30,20 +30,24 @@ std::vector<std::shared_ptr<AbstractTask>> OperatorTask::make_tasks_from_operato
 std::shared_ptr<AbstractTask> OperatorTask::_add_tasks_from_operator(
     const std::shared_ptr<AbstractOperator>& op, std::vector<std::shared_ptr<AbstractTask>>& tasks,
     std::unordered_map<std::shared_ptr<AbstractOperator>, std::shared_ptr<AbstractTask>>& task_by_op) {
+  // Early out: We do not want to re-run operators, if they have already been executed by previous tasks.
+  if (op->executed()) return nullptr;
```

Comment:
Is this path taken anywhere? As `executed` may change while this method is called, it is somewhat unpredicable in a multi-threaded environment if this method returns nullptr or not, isn't it? Even if it was predictable, it may be an unexpected return value.

Reply:
> Is this path taken anywhere?

In some cases, it will be taken by the ExpressionEvaluator, which calls `OperatorTask::make_tasks_from_operator` at runtime. Some other operators might already have executed the subquery or parts of it.

>  As executed may change while this method is called, it is somewhat unpredicable in a multi-threaded environment if this method returns nullptr or not, isn't it?

`op` is about to be executed, so why should we end up with a nullptr?

Reply:
More as a note for myself - I'll have to think about this tomorrow: Can two ExpressionEvaluators that decide to execute the PQP at the very same moment at the very same speed result in the operator being executed twice? We can't expect safety from the scheduler system, because we create independent tasks. Do we want to atomically prohibit multiple tasks being created from an operator?

Reply:
Since `AbstractOperator::execute` has the following in place,

```cpp
if (_executed) return;
```

we should not run into problems. However, we should consider using atomics for our flag:

```cpp
 bool _executed = false;
```

Reply:
Right but _executed is set once the operator is finished. Couldn't the two task both decide that they should start the task?

> However, we should consider using atomics for our flag

That would help only partially, as `_executed` is set at the end. We could have an atomic `_started` (or, better, a `_state = [Created, Running, Executed, Cleared]` flag. Also remember that calling `if` on an atomic is close to pointless. We would need an atomic exchange.

Reply:
> We could have an atomic _started (or, better, a _state = [Created, Running, Executed, Cleared] flag.

Sounds good, but we would need additional synchronization logic.
For example, when a task starts and sees an operator in the `Running` state, it should sleep and wait for the results. Later, the operator should ping waiting tasks after it has changed to `Executed`.

> Also remember that calling if on an atomic is close to pointless. We would need an atomic exchange.

As as the Doc implies, shouldn't Reads be synchronized by default?

> If one thread writes to an atomic object while another thread reads from it, the behavior is well-defined (see memory model for details on data races).
> In addition, accesses to atomic objects may establish inter-thread synchronization and order non-atomic memory accesses as specified by std::memory_order.

Reply:
>  For example, when a task starts and sees an operator in the Running state, it should sleep and wait for the results. Later, the operator should ping waiting tasks after it has changed to Executed.

Brrr. That sounds like the job of the OperatorTask. Are we sure that we need to allow multiple OperatorTasks per operator? That sounds like it should not happen - even if the current design makes it possible. In the worst case (!), we could have a weak_ptr from the operator to its OperatorTask. Not sure how much change this would require.

> As as the Doc implies, shouldn't Reads be synchronized by default

Yes, but only the reads are. In the time between reading the value and evaluating the if clause, everything might have changed again: https://stackoverflow.com/questions/26656823/basic-usage-of-conditionals-with-stdatomict

Reply:
Thanks. Got it.

> Brrr. That sounds like the job of the OperatorTask. Are we sure that we need to allow multiple OperatorTasks per operator? That sounds like it should not happen - even if the current design makes it possible. In the worst case (!), we could have a weak_ptr from the operator to its OperatorTask. Not sure how much change this would require.

I think it is a rare edge case to see an operator in the `Running`-stage when calling `execute()`. 
We should add operator states, as you described, for transparency and add Asserts for rare edge cases, so we get notified when things go wrong badly.

Reply:
Rare, but possible, correct? As such, we should handle it in a way that does not involve failing. However, in the interest of moving this PR forward, I would be fine with identifying these cases for now and solving the issue in a follow-up PR.

Reply:
Good. I'll continue with this on Thursday.